### ISSUE TITLE
test vgg-face with 4096 dimensional outputs on LFW dataset

### ISSUE THREAD

**Issue by serengil** [serengil/deepface#944]
Currently, our vgg-face model has 2622 outputs. But this is the output for classification model of vgg-face model. In original paper, they are using its early layer with 4096 dimensions for descriptor. Test this and compare with exisiting model.

**Comment by serengil** [comment]
If we chang, VGGFace.py as 

```python
    # from
    # vgg_face_descriptor = Model(inputs=model.layers[0].input, outputs=model.layers[-2].output)
    # to
    base_model_output = Sequential()
    base_model_output = Convolution2D(4096, (1, 1), name="predictions")(model.layers[-4].output)
    base_model_output = Flatten()(base_model_output)
    vgg_face_descriptor = Model(inputs=model.input, outputs=base_model_output)
```

and update distance.py as 

```json
"VGG-Face": {"cosine": 0.66, "euclidean": 0.61, "euclidean_l2": 1.12}
```

cosine and euclidean_l2 metrics are giving good results but euclidean is giving bad result as shown in attachment.

<img width="1362" alt="early-results" src="https://github.com/serengil/deepface/assets/18491038/6fd22d43-75ee-48ab-833f-9be20517a0bd">

**Comment by serengil** [comment]
on the other hand, model with 2622 dimensional vectors has the following distributions. euclidean is more stable, but cosine and euclidean l2 was more stable in 4096d one.

<img width="970" alt="results-2622d" src="https://github.com/serengil/deepface/assets/18491038/ad234389-9df2-4569-a0ac-4c1a45544377">

**Comment by serengil** [comment]
Finally, if we replace the vgg descriptor as 

```python
    base_model_output = Sequential()
    base_model_output = Convolution2D(4096, (1, 1), name="predictions")(model.layers[-4].output)
    base_model_output = Flatten()(base_model_output)
    base_model_output = Activation("softmax")(base_model_output)
    vgg_face_descriptor = Model(inputs=model.input, outputs=base_model_output)
```

(Just added softmax layer in addition to first experiment)

We have these results:

<img width="777" alt="4096-with-softmax" src="https://github.com/serengil/deepface/assets/18491038/4163c59e-a030-4b3c-b1bf-1ceb89d526da">

So, dropping softmax layer makes it more stable.

**Comment by serengil** [comment]
- [ ] try 4096 dimensional model without softmax layer on LFW dataset.
- [ ] if cosine and euclidean_l2 overperforms with this form, then overwrite euclidean with euclidean_l2 for vgg-face model.

**Comment by serengil** [comment]
When I do perform a test for VGG-Face and mtcnn detector

| Model | Distance Metric | LFW Score |
| 2622d | euclidean_l2 | 90.3 |
| 2622d | cosine | 90.3 |
| 4096d | euclidean_l2 | 96.0 |
| 4096d | cosine | 96.0 |

So, using its early layer obviously contributed the score

**Comment by serengil** [comment]
To avoid the failure in euclidean distance metric, what if we add normalization layer to the descriptor?

```python
base_model_output = Sequential()
    base_model_output = Convolution2D(4096, (1, 1), name="predictions")(model.layers[-4].output)
    base_model_output = Flatten()(base_model_output)
    base_model_output = Lambda(lambda x: K.l2_normalize(x, axis=1), name="norm_layer")(
        base_model_output
    )
    vgg_face_descriptor = Model(inputs=model.input, outputs=base_model_output)
```

distribution graph on deepface unit test items according to [this tutorial](https://sefiks.com/2020/05/22/fine-tuning-the-threshold-in-face-recognition/) becomes stable for euclidean (because it would be equivalent to euclidean_l2), and negative impact on cosine and euclidean_l2.
<img width="777" alt="norm_layer" src="https://github.com/serengil/deepface/assets/18491038/668b71d7-ef26-46a2-8a6c-c719aaf1d31d">

**Comment by serengil** [comment]
To sum up, (model_name ="VGG-Face", detector_backend = "mtcnn", align = True, enforce_detection = False) configuration is tested on LFW dataset's test samples (1000 instances distributed 50% 50% for same person and different persons); 

- In the 4096d one, there is **NO** softmax layer and there is a l2 normalization layer.

| distance metric | 2622d | 4096d |
| --- | --- | --- |
| cosine | 90.3 | 96.0 |
| euclidean | 82.2 | 96.0 |
| euclidean_l2 | 90.3 | 96.0 |

and these are the thresholds:

cosine 0.7077406336745886
euclidean 1.1897399682146557
euclidean_l2 1.1897399998945895

So, doing this action increases the model performance 6% o 14%!

**Comment by serengil** [comment]
now, it is creating different embeddings in every run. we should get rid of the Convolution2D in the descriptor because it generates random weights for that layer!

```python
base_model_output = Sequential()
    base_model_output = Flatten()(model.layers[-5].output)
    base_model_output = Lambda(lambda x: K.l2_normalize(x, axis=1), name="norm_layer")(
        base_model_output
    )
    vgg_face_descriptor = Model(inputs=model.input, outputs=base_model_output)
```

**Comment by serengil** [comment]
With the new structure, same accuracy values gotten

| distance metric | 2622d | 4096d |
| --- | --- | --- |
| cosine | 90.3 | 96.0 |
| euclidean | 82.2 | 96.0 |
| euclidean_l2 | 90.3 | 96.0 |

for these thresholds

| metric | threshold |
| --- | --- |
| cosine | 0.6848793279466694 |
| euclidean | 1.1735069839807788 |
| euclidean_l2 | 1.1735069853005953 |

**Comment by serengil** [comment]
Closed with PR - https://github.com/serengil/deepface/pull/948

**Comment by serengil** [comment]
vgg-face is the default facial recognition model. so, this improvement will affect every user of deepface.
